Ajout de l'interface de promotion

Eskimon opened the pull request:
| Q | R |
| --- | --- |
| Correction de bugs ? | [oui] |
| Nouvelle Fonctionnalité ? | [oui] |
| Tickets concernés | #54 (yeah !!) |

Cette PR se propose d'apporter un nouveau moyen pour upgrader un utilisateur en staff ou en super-utilisateur via la fiche membre
#### QA
- Seul un super user (compte 'admin' donc) doit pouvoir voir le menu "Promouvoir" sur la fiche d'un utilisateur
- Un lambda ou staff ne doit pas pouvoir changer quoi que ce soit s'il voulait (bidouiller le template si vous voulez voir apparaitre le menu)
- ATTENTION, les fixtures sont mauvaises, pour vérifier vos changements utiliser l'interface /admin pour regarder l'impact

INFO : Les TU arriveront dans la PR que j'ai en cours à ce propos !
# Ne pas merger, je vais refaire en plus propre

(mais si vous mergez quand meme c'est pas dramatique car ca marche)

Eskimon committed:
Ajout de l'interface de promotion

pierre-24 commented:
Rapport de QA : 
- Seul un superuser peut changer les droits : OK
- Les droits sont effectivement bien changés : OK (à checker en préprod' que ça a un réel effet pour "satff")
- Le changement de droit a un effet immédiat sur le compte de l'utilisateur. Dans le sens de la promotion, ça va. Dans l'autre sens, comportement indéfini sur certaines parties du site. J'ai rien pris en défaut quand j'ai testé (en gros, on se mange une 403 si on fait une action de staff après perte de droit, comme prévu), mais on peut pas l'exclure. Mais OK.
- On peut changer ces propres droits, ce qui signifie qu'un admin pourrait se rétrograder. Bug ou fonctionnalité ?

Bref, **merge OK**, après rélfexion sur les remarques ;)

Eskimon commented:
La 403 me parait tout ce qu'il y a de plus normal. C'est bien un code de "droit insuffisant" comme il faut.

> On peut changer ces propres droits, ce qui signifie qu'un admin pourrait se rétrograder. Bug ou fonctionnalité ?

J'avoue avoir pris la décision dans mon coin sans trop y réfléchir en me disant "pourquoi pas". Si vraiment c'est gênant c'est facile à changer pour qu'un user ne puisse pas se rétrograder.

pierre-24 commented:
Dans ce cas, merge accordé :)

Eskimon commented:
Je tiens à rappeler : *_Un test sera nécessaire en pré-prod pour s'assurer de l'action "devenir staff" *_

Eskimon commented on templates/member/profile.html:
```diff
@@ -169,6 +169,26 @@ <h2 class="ico-after ico-articles">Derniers articles</h2>
+                                Quelles sont les droit de ce membre ?
+                            </p>
+                            {% with status=usr|state %}
+                                <label class="checkbox"><input type="checkbox" name="staff-check" value="isstaff" {% for group in usr.groups.all %}{% if 'staff' == group.name %} checked="checked" {% endif %}{% endfor %}>Staff</label>
```
Il y a t-il n moyen de faire plus propre qu'une boucle ? (un genre de "key()" ?)

gustavi replied:
Non, pas avec du template pur.

Eskimon commented:
J'ai aussi une question sur un bout de front dont je ne suis pas sur si c'est bien opti (mais seul les superuser le verront tfacon)

Eskimon committed:
Correction d'uen faute

Eskimon committed:
Ajout d'un test et de messages

Un super utilisateur ne peut plus se supprimer des super user
Des messages d'infos sur l'action ont été rajouté

Eskimon commented:
désolé @pierre-24 j'ai fait une petite modif pour ajouter des messages et faire en sorte qu'un super user ne pourra pas se retirer des super user (évite les fausses manips lorsqu'on pense etre sur un profil alros qu'on est sur le sien par exemple). Tu peux re-QAitser

pierre-24 commented:
Pas de soucis, @Eskimon, c'est mieux maintenant que dans une PR en urgence.

Rapport de QA. OK, on a bien le message "un super user ne peut pas se retirer ces droits de super-utilisateur". Par contre, les droits de "staff" sont bien retirés ("<user> n'est maintenant plus staff").

Eskimon commented:
> Par contre, les droits de "staff" sont bien retirés (" n'est maintenant plus staff").

C'est normal, ca me semble deux faits indépendants (on peut etre super user pour le site sans vouloir etre modo)

pierre-24 commented:
OK, bah merge ok dès que Travis est fini :)

firm1 commented:
Attention : pensez a unfollow tous les topics suivi par quelqu'un qu'on
retrograde pour pas continuer a recevoir des notifs.
Le 29 juil. 2014 20:52, "Pierre Beaujean" notifications@github.com a
écrit :

> OK, bah merge ok dès que Travis est fini :)
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/zestedesavoir/zds-site/pull/1320#issuecomment-50521147
> .

pierre-24 commented:
Ah; Zut. Bon, pas merge tout de suite

Eskimon commented:
> Attention : pensez a unfollow tous les topics suivi par quelqu'un qu'on
> retrograde pour pas continuer a recevoir des notifs.

Bien vu ca !! Je n'y avais carrément plus pensé !

Eskimon committed:
Correction + unfollow des topics staff

Eskimon commented:
Voila j'ai rajouté tout ca :
- Correction sur le limeogage d'un super_user (j'avais pas fait gaffe mais un superuser pouvait plus du tout supprimer un superuser en fait)
- Unfollow les sujets staff suivis

~~Attention, je viens de me rendre compte qu'un bout de code n'est pas a sa place, je vais corriger et tenter un rebase...~~ C'est bon :)

Eskimon committed:
Ajout d'un message

Eskimon commented:
A ne pas merger, j'ai oublie que le group "developpeur" devait aussi etre assigne via cette interface !

firm1 commented:
> A ne pas merger, j'ai oublie que le group "developpeur" devait aussi etre assigne via cette interface !

En lisant ça je suis aller checker le code, et mauvaise surprise, le nom du group "staff" y est en dur. ça pose un problème, celui de bloquer le nom du groupe, et donc pour chaque nouveau groupe, être obligé de rajouter une condition dans le code.

Pour le faire de manière propre, il faudrait plutot : 
1- creer un formulaire dans le form.py du membre avec un seul champ group qui affiche la liste des groupes. ça ressemblera a un truc du style.

``` python
class PromoteMemberForm(forms.Form):

    group = forms.ModelChoiceField(
        label="Groupe de promotion",
        queryset=Groupe.objects.all(),
        required=False,
    )

    def __init__(self, *args, **kwargs):
        super(PromoteMemberForm, self).__init__(*args, **kwargs)
        self.helper = FormHelper()
        self.helper.form_class = 'content-wrapper'
        self.helper.form_method = 'post'

        self.helper.layout = Layout(
            Field('group'),
            StrictButton('Valider', type='submit'),
        )
```

2- modifier la vue `upgrade_profile` pour qu'elle se charge de rajouter et d'enlever un groupe à un membre.

Eskimon commented:
Ok, je regarderais ca ce soir... Je repartirais surement sur une PR vierge pour avoir des commits plus propre et un ensemble plus zoli

gustavi commented:
Sinon on peut stocker le nom du groupe staff dans les settings. S'il est à None on crée un groupe « Staff » pour le dev.

Mais la solution de @firm1 est propre.

Eskimon commented:
Je ferme ici, la suite se passe la ba : https://github.com/zestedesavoir/zds-site/pull/1344
